s3/pkg/datastore/yig/meta/db/tidb/multiparts.go

```diff
@@ -0,0 +1,145 @@
+	log "github.com/sirupsen/logrus"
+)
+
+/*
+* ListParts: list all the parts which belong to the uploadId.
+* @return: the related parts which are sorted by PartNum asc.
+*
+ */
+func (t *Tidb) ListParts(uploadId uint64) ([]*types.PartInfo, error) {
+	sqlText := "select upload_id, part_num, object_id, location, pool, offset, size, etag, flag, create_time, update_time from multiparts where upload_id = ?"
+	var parts []*types.PartInfo
+	rows, err := t.DB.Query(sqlText, uploadId)
+	if err != nil {
+		log.Errorf("failed to query parts for uploadId %d, err: %v", uploadId, err)
+		return nil, err
+	}
+	defer rows.Close()
+	for rows.Next() {
+		part := &types.PartInfo{}
+		var createTime sql.NullString
+		var updateTime sql.NullString
+		err = rows.Scan(&part.UploadId, &part.PartNum, &part.ObjectId, &part.Location, &part.Pool, &part.Offset, &part.Size, &part.Etag, &part.Flag, &createTime, &updateTime)
+		if err != nil {
+			log.Errorf("failed to scan rows for uploadId %d, err: %v", uploadId, err)
+			return nil, err
+		}
+		if createTime.Valid {
+			part.CreateTime, err = time.Parse(mtypes.TIME_LAYOUT_TIDB, createTime.String)
+			if err != nil {
+				log.Errorf("failed to parse create_time: %s, err: %v", createTime.String, err)
+				return nil, err
+			}
+		}
+		if updateTime.Valid {
+			part.UpdateTime, err = time.Parse(mtypes.TIME_LAYOUT_TIDB, updateTime.String)
+			if err != nil {
+				log.Errorf("failed to parse update_time: %s, err: %v", updateTime.String, err)
+				return nil, err
+			}
+		}
+		parts = append(parts, part)
+	}
+	err = rows.Err()
+	if err != nil {
+		log.Errorf("failed to iterate the rows for uploadId %d, err: %v", uploadId, err)
+		return nil, err
+	}
+	sort.Sort(types.ByPartNum(parts))
+	return parts, nil
+}
+
+func (t *Tidb) PutPart(partInfo *types.PartInfo) error {
+	sqlText := "insert into multiparts(upload_id, part_num, object_id, location, pool, offset, size, etag, flag) values(?, ?, ?, ?, ?, ?, ?, ?, ?)"
+	var err error
+	var tx *sql.Tx
+	tx, err = t.DB.Begin()
+	if err != nil {
+		log.Errorf("failed to Begin a transaction for %v, err: %v", partInfo, err)
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			// do not use the err since since by doing so, it will overwite the original error.
+			if rErr := tx.Rollback(); rErr != nil {
+				log.Errorf("perform rollback for partInfo(%v) failed with err: %v", partInfo, rErr)
+			}
+		} else {
+			// should check whether the transaction commit succeeds or not.
+			if err = tx.Commit(); err != nil {
+				log.Errorf("perform commit for partInfo(%v) failed with err: %v", partInfo, err)
+			}
+		}
+	}()
+
+	_, err = tx.Exec(sqlText, partInfo.UploadId, partInfo.PartNum, partInfo.ObjectId, partInfo.Location, partInfo.Pool, partInfo.Offset, partInfo.Size, partInfo.Etag, partInfo.Flag)
+	if err != nil {
+		log.Errorf("failed to save partInfo(%v), err: %v", partInfo, err)
+		return err
+	}
+	// must return err instead of nil, because commit may return error.
+	return err
```

Comment:
The error of tx.Commit() cannot be returned.

Reply:
@sfzeng , when error happens if we call tx.Commit(), that error will be returned to the caller, and  it will not be lost. the scope of err variable is in that whole function.

Reply:
Hi @hopkings2008 I tried the following program and got "test error!", please check:
package main

import (
        "fmt"
        "errors"
)

func test() error {
        var err error
        defer func() {
                err = errors.New("defer error!")
        }()

        err = errors.New("test error!")
        return err
}

func main() {
        err := test()
        fmt.Printf("err=%v\n", err)
}

Reply:
thanks very much @sfzeng for clarifying the error made by my original logic to return error. i have changed the logic of returning error by using named return value. pls help to check it again, thanks a lot.